src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncMultipleVariantsIT.java

```diff
@@ -0,0 +1,117 @@
+import static java.util.Collections.singletonList;
+
+import com.commercetools.sync.commons.asserts.statistics.AssertionsForStatistics;
+import com.commercetools.sync.commons.exceptions.SyncException;
+import com.commercetools.sync.commons.utils.TriConsumer;
+import com.commercetools.sync.products.ProductSync;
+import com.commercetools.sync.products.ProductSyncOptions;
+import com.commercetools.sync.products.ProductSyncOptionsBuilder;
+import com.commercetools.sync.products.helpers.ProductSyncStatistics;
+import io.sphere.sdk.products.ProductDraft;
+import io.sphere.sdk.products.ProductProjection;
+import io.sphere.sdk.products.commands.ProductCreateCommand;
+import io.sphere.sdk.producttypes.ProductType;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class ProductSyncMultipleVariantsIT {
+
+  private static ProductType productType;
+
+  private List<String> errorCallBackMessages;
+
+  private List<String> warningCallBackMessages;
+
+  private List<Throwable> errorCallBackExceptions;
+
+  private ProductSyncOptions syncOptions;
+
+  /**
+   * Delete all product related test data from the target project. Then creates for the target CTP
+   * project price a product type, a tax category, 2 categories, custom types for the categories and
+   * a product state.
+   */
+  @BeforeAll
+  static void setup() {
+    deleteProductSyncTestData(CTP_TARGET_CLIENT);
+
+    productType = createProductType(PRODUCT_TYPE_RESOURCE_PATH, CTP_TARGET_CLIENT);
+  }
+
+  /**
+   * Deletes Products and Types from the target CTP project, then it populates target CTP project
+   * with product test data.
+   */
+  @BeforeEach
+  void setupTest() {
+    clearSyncTestCollections();
+    deleteAllProducts(CTP_TARGET_CLIENT);
+    syncOptions = buildSyncOptions();
+
+    final ProductDraft productDraft =
+        createProductDraftBuilder(
+                PRODUCT_KEY_1_MULTIPLE_VARIANTS_RESOURCE_PATH, productType.toReference())
+            .build();
+
+    executeBlocking(CTP_TARGET_CLIENT.execute(ProductCreateCommand.of(productDraft)));
+  }
+
+  @AfterAll
+  static void tearDown() {
+    deleteProductSyncTestData(CTP_TARGET_CLIENT);
+  }
+
+  @Test
+  void sync_withReassignmentOfMasterVariant_shouldSyncProductCorrectly() {
+    // preparation
+    final ProductDraft productDraft =
+        createProductDraftBuilder(
+                PRODUCT_KEY_2_MULTIPLE_VARIANTS_RESOURCE_PATH, productType.toReference())
+            .build();
+
+    final ProductSync productSync = new ProductSync(syncOptions);
+    final ProductSyncStatistics syncStatistics =
+        executeBlocking(productSync.sync(singletonList(productDraft)));
+
+    AssertionsForStatistics.assertThat(syncStatistics).hasValues(1, 0, 1, 0, 0);
```

Comment:
Check the attributes / sku of master variant here.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/918/commits/7912b61cfb31fd078bab2032e6bb9206ae525950